File: algorithms/2_medium/M/01898_maximum_number_of_removable_characters/binary_search.cpp
class Solution {
    inline static bool is_ok(
        span<const char> s,
        span<const char> p,
        span<const char> can_use
    ) {
        const size_t s_len = s.size();
        const size_t p_len = p.size();
        size_t s_idx = 0;
        size_t p_idx = 0;
        while (s_idx < s_len && p_idx < p_len) {
            if (can_use[s_idx] != 0 && s[s_idx] == p[p_idx]) {
                p_idx += 1;
            }

            s_idx += 1;
        }

        return p_idx == p_len;
    }
public:
    static int maximumRemovals(const string& s, const string& p, const vector<int>& removable) {
        ptrdiff_t lo = 0;
        ptrdiff_t hi = std::ssize(removable) - 1;
        auto can_use = vector<char>(s.size());

        while (lo <= hi) {
            const ptrdiff_t mid = lo + (hi - lo) / 2;

            for (auto& c : can_use) { c = 1; }

            for (ptrdiff_t i = 0; i <= mid; i += 1) {
                can_use[removable[i]] = 0;
            }

            if (is_ok(s, p, can_use)) {
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }

        return static_cast<int>(hi + 1);
    }
};

File: algorithms/2_medium/M/01898_maximum_number_of_removable_characters/binary_search.rs
impl Solution {
    #[inline(always)]
    fn is_ok(s: &[u8], p: &[u8], can_use: &[bool]) -> bool {
        let (s_len, p_len) = (s.len(), p.len());
        let (mut s_idx, mut p_idx) = (0_usize, 0_usize);
        while s_idx < s_len && p_idx < p_len {
            if can_use[s_idx] && s[s_idx] == p[p_idx] {
                p_idx += 1;
            }

            s_idx += 1;
        }

        p_idx == p_len
    }

    pub fn maximum_removals(s: String, p: String, removable: Vec<i32>) -> i32 {
        let (s_bytes, p_bytes) = (s.as_bytes(), p.as_bytes());
        let (mut lo, mut hi) = (0_isize, removable.len() as isize - 1);
        let mut can_use = vec![true; s.len()];
        while lo <= hi {
            let mid = lo + (hi - lo) / 2;
            can_use.iter_mut().for_each(|x| *x = true);
            for i in 0..=mid {
                can_use[removable[i as usize] as usize] = false;
            }

            if Self::is_ok(s_bytes, p_bytes, &can_use) {
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }

        (hi + 1).try_into().unwrap()
    }
}
